### ISSUE TITLE
[MINSTALL-82] Add sha 256 support

### ISSUE THREAD

**Issue by jira-importer** [apache/maven-install-plugin#301]
**[Todd Nine](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=todd@spidertracks.co.nz)** opened **[MINSTALL-82](https://issues.apache.org/jira/browse/MINSTALL-82?redirect=false)** and commented

Currently no facility for sha 256 exists in the install plugin.  Utilities such as chef depend on sha 256 to validate downloaded files.  I have added the functionality to the plugin.  Node that my 2 plexus classes are not properly loaded and I have hard coded the sha256 implementation.  I'm unfamiliar with plexus, but all the classes should be correct to submit to the digest utilities.

Thanks,
Todd

---

**Affects:** 2.3.1

**Attachments:**
- [maven-install-sha3.patch](https://issues.apache.org/jira/secure/attachment/12782937/maven-install-sha3.patch) (_6.93 kB_)
- [MINSTALL_82_add__sha256_files__based_on_trunk_commit_1724329_.patch](https://issues.apache.org/jira/secure/attachment/12782781/MINSTALL_82_add__sha256_files__based_on_trunk_commit_1724329_.patch) (_11.35 kB_)
- [sha256.patch](https://issues.apache.org/jira/secure/attachment/12716600/sha256.patch) (_6.78 kB_)

**Issue Links:**
- [MINSTALL-143](https://issues.apache.org/jira/browse/MINSTALL-143) Move checksum generation from install to deploy plugin
 (_**"is blocked by"**_)
- [MINSTALL-138](https://issues.apache.org/jira/browse/MINSTALL-138) option to calculate more checksum such sha-256 sha-512
 (_**"Is contained by"**_)
- [MINSTALL-134](https://issues.apache.org/jira/browse/MINSTALL-134) Remove checksum generation

- [MRESOLVER-56](https://issues.apache.org/jira/browse/MRESOLVER-56) Support SHA-256 and SHA-512 as checksums
 (_**"is superceded by"**_)

4 votes, 8 watchers

**Comment by jira-importer** [comment]
**[John Dyer](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=johntdyer)** commented

This would be GREAT!

**Comment by jira-importer** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

Would be nice this could be fixed at [plexus-digest](http://plexus.codehaus.org/plexus-components/plexus-digest). We could poke some other Maven teammembers which do have commit rights on that project if there's a patch.

**Comment by jira-importer** [comment]
**[Sven Bunge](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=bigd)** commented

This issue is 5 years old but still up-to-date. The NIST and other organisations suggest to switch to SHA-256 immediatly to prevent collision attacks. I think maven artefacts should add .sha256 hash files to their artifacts / releases as well -- then maven is able to check them if they exists.

The plexus-digest has been moved to codehaus but seems not under continuous development. I've raised a [Pull Request](https://github.com/codehaus-plexus/plexus-digest/pull/2) on their repository. But nevertheless: Noone of this digesters is used in this install plugin yet. So let's move forward and switch to the plexus-digest plugin later on.

I've attached a second sha256-patch -- you can integrate mine if the old one isn't not applicable due the age.

**Comment by jira-importer** [comment]
**[Alexander Kjäll](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=capitol)** commented

I thought about this yesterday and wrote a patch myself, and now i see that someone else had the exact same idea :D

I'll attach my patch for adding SHA3 support, so that you have something to choose between.

**Comment by jira-importer** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

We're working on version 3.0 of the plugin, which will use Aether. I thought that Aether is capable to generate such hashes as well. In that case the duplicate code will be removed from this plugin. I hope that Aether is pluggable enough to add extra hashgenerators, something you can configure from the plugin. I don't think all these hashes should be generated by default, that's up to the user.

**Comment by jira-importer** [comment]
**[Alexander Kjäll](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=capitol)** commented

[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte) That sounds reasonable. Would you merge a patch that generated warnings when insecure hashes was being generated in order to inform the user that the hash is broken?

**Comment by jira-importer** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

Insecure or broken? Anyway, even though these files are now generated by the maven-install-plugin, I think they belong to the maven-deploy-plugin. And this matches the behavior I've seen by Aether. I think it is the responsibility of the artifact repository manager to define the policy on how strong hashes should be, so I have my doubts if warnings should be generated by plugins.

**Comment by jira-importer** [comment]
**[Konrad Windszus](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kwin)** commented

Now that we have MRESOLVER-56 we can probably close this one as won't fix.
